### ISSUE TITLE
Avoid wrapping mcols<- with DataFrame() when replacing NULL

### ISSUE THREAD

**Issue by LTLA** [Bioconductor/S4Vectors#52]
`mcols<-` seems to wrap things in a `DataFrame()` call when it's replacing a `NULL` value:

```r
library(S4Vectors)
X <- List(A=1, B=2, C=3)
mcols(X)$thing <- DataFrame(B=LETTERS[1:3])
mcols(X)
## DataFrame with 3 rows and 1 column
##       thing.B
##   <character>
## A           A
## B           B
## C           C
```

This is undesirable as `mcols(X)$thing` doesn't give me what I put in.

<details>
<summary>Session info</summary>

```
R version 3.6.0 Patched (2019-05-02 r76458)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.3 LTS

Matrix products: default
BLAS:   /home/luna/Software/R/R-3-6-branch-dev/lib/libRblas.so
LAPACK: /home/luna/Software/R/R-3-6-branch-dev/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
[1] S4Vectors_0.23.23   BiocGenerics_0.31.6

loaded via a namespace (and not attached):
[1] compiler_3.6.0
```
</details>

**Comment by hpages** [comment]
Hmm, this actually reveals a broader problem with how the `mcols()` setter handles a right value supplied as a list:
```
library(S4Vectors)
X <- List(A=1, B=2, C=3)
mcols(X) <- list(stuff=DataFrame(aa=11:13, bb=21:23), more_stuff=as.list(31:33))
mcols(X)
# DataFrame with 3 rows and 5 columns
#    stuff.aa  stuff.bb more_stuff.X31L more_stuff.X32L more_stuff.X33L
#   <integer> <integer>       <integer>       <integer>       <integer>
# A        11        21              31              32              33
# B        12        22              31              32              33
# C        13        23              31              32              33
```
Not good!

Should be addressed in S4Vectors 0.23.24 (commit 05a99cd0eb1e27373ece2c28c75dd0642a070016):
```
library(S4Vectors)
X <- List(A=1, B=2, C=3)
mcols(X)$thing <- DataFrame(B=LETTERS[1:3])
mcols(X)
# DataFrame with 3 rows and 1 column
#         thing
#   <DataFrame>
# A           A
# B           B
# C           C

mcols(X) <- list(stuff=DataFrame(aa=11:13, bb=21:23), more_stuff=as.list(31:33))
mcols(X)
# DataFrame with 3 rows and 2 columns
#         stuff more_stuff
#   <DataFrame>     <list>
# A       11:21         31
# B       12:22         32
# C       13:23         33
mcols(X)$stuff
# DataFrame with 3 rows and 2 columns
#          aa        bb
#   <integer> <integer>
# 1        11        21
# 2        12        22
# 3        13        23
mcols(X)$more_stuff
# [[1]]
# [1] 31
# 
# [[2]]
# [1] 32
# 
# [[3]]
# [1] 33
```

Thanks for the catch!

**Comment by hpages** [comment]
Closing this.
